File: 02_Go/incoming/behaviour.go
package incoming

import "fmt"

type Behaviour uint8

const (
	BehaviourNone Behaviour = 0

	BehaviourHighPriority Behaviour = 1 << (iota - 1)
	BehaviourLongRunning
	BehaviourResourceIntensive
	BehaviourRequiresAffinity
	BehaviourRetryable
	BehaviourExactlyOnce
)

func (b Behaviour) HasFlag(other Behaviour) bool {
	return b&other == other
}

func (b Behaviour) String() string {
	switch b {
	case BehaviourNone:
		return "None"
	case BehaviourHighPriority:
		return "HighPriority"
	case BehaviourLongRunning:
		return "LongRunning"
	case BehaviourResourceIntensive:
		return "ResourceIntensive"
	case BehaviourRequiresAffinity:
		return "RequiresAffinity"
	case BehaviourRetryable:
		return "Retryable"
	case BehaviourExactlyOnce:
		return "ExactlyOnce"
	default:
		return fmt.Sprintf("%d", int(b))
	}
}

var allBehaviours = []Behaviour{
	BehaviourHighPriority,
	BehaviourLongRunning,
	BehaviourResourceIntensive,
	BehaviourRequiresAffinity,
	BehaviourRetryable,
	BehaviourExactlyOnce,
}

func GetBehaviours() []Behaviour {
	return allBehaviours
}

File: 06_Kotlin/workers/src/main/kotlin/incoming/Behaviour.kt
package org.andreastme.langpermonth.incoming

@JvmInline
value class Behaviour private constructor(val value: Byte) {

    infix fun or(other: Behaviour): Behaviour =
        Behaviour((value.toInt() or other.value.toInt()).toByte())

    fun hasFlag(flag: Behaviour): Boolean =
        (value.toInt() and flag.value.toInt()) != 0

    companion object {
        val None = Behaviour(0)
        val HighPriority = Behaviour(1 shl 0)
        val LongRunning = Behaviour(1 shl 1)
        val ResourceIntensive = Behaviour(1 shl 2)
        val RequiresAffinity = Behaviour(1 shl 3)
        val Retryable = Behaviour(1 shl 4)
        val ExactlyOnce = Behaviour(1 shl 5)

        val entries = setOf<Behaviour>(
            None,
            HighPriority,
            LongRunning,
            ResourceIntensive,
            RequiresAffinity,
            Retryable,
            ExactlyOnce
        )

        fun from(value: Byte): Behaviour = Behaviour(value)
    }

    private constructor(value: Int) : this(value.toByte())
}